lib/smart_proxy_container_gateway/container_gateway_main.rb

```diff
@@ -57,140 +59,124 @@ def tags(repository, params = {})
         return false if repository.nil?
 
         # Repository doesn't require auth
         return true unless repository.auth_required
 
-        if username && user_token_is_valid && repository.auth_required
+        if username && user_token_is_valid
           # User is logged in and has access to the repository
-          user = User.find(name: username)
-          return !user.repositories_dataset.where(name: repo_name).first.nil?
+          return database[:repositories_users].where(repository_id: repository.id, user_id: database[:user].select(:id).where(name: username)).exists
         end
 
         false
       end
 
       def token_user(token)
-        User[AuthenticationToken.find(token_checksum: Digest::SHA256.hexdigest(token)).user_id]
+        database[:users].find(id: database[:authentication_tokens].where(token_checksum: checksum(token)).select(:user_id))
       end
 
       def valid_token?(token)
-        AuthenticationToken.where(token_checksum: Digest::SHA256.hexdigest(token)).where do
+        database[:authentication_tokens].where(token_checksum: checksum(token)).where do
           expire_at > Sequel::CURRENT_TIMESTAMP
-        end.count.positive?
+        end.exists
       end
 
       def insert_token(username, token, expire_at_string, clear_expired_tokens: true)
-        checksum = Digest::SHA256.hexdigest(token)
-        user = User.find_or_create(name: username)
-
-        AuthenticationToken.where(:token_checksum => checksum).delete
-        AuthenticationToken.create(token_checksum: checksum, expire_at: expire_at_string.to_s, user_id: user.id)
-        AuthenticationToken.where { expire_at < Sequel::CURRENT_TIMESTAMP }.delete if clear_expired_tokens
-      end
-
-      def initialize_db
-        file_path = Proxy::ContainerGateway::Plugin.settings.sqlite_db_path
-        sqlite_timeout = Proxy::ContainerGateway::Plugin.settings.sqlite_timeout
-        conn = Sequel.connect("sqlite://#{file_path}", timeout: sqlite_timeout)
-        container_gateway_path = $LOAD_PATH.detect { |path| path.include? 'smart_proxy_container_gateway' }
-        begin
-          Sequel::Migrator.check_current(conn, "#{container_gateway_path}/smart_proxy_container_gateway/sequel_migrations")
-        rescue Sequel::Migrator::NotCurrentError
-          migrate_db(conn, container_gateway_path)
-        end
-        conn
-      end
+        user = database[:users].find_or_create(name: username)
 
-      private
+        authentication_tokens = database[:authentication_tokens]
+
+        # TODO: check if this upsert is actually correct
+        authentication_tokens.insert_conflict(:update).insert(token_checksum: checksum(token), expire_at: expire_at_string.to_s, user_id: user.id)
 
-      def migrate_db(db_connection, container_gateway_path)
-        Sequel::Migrator.run(db_connection, "#{container_gateway_path}/smart_proxy_container_gateway/sequel_migrations")
+        # TODO: create a background service that does this
```

Comment:
You can start a service within smart-proxy.

For example: https://github.com/theforeman/smart-proxy/blob/e8fe59f83843ca619264e8d4757bc99fcde93fa9/modules/dhcp_isc/dhcp_isc_plugin.rb#L18

This is configured via: https://github.com/theforeman/smart-proxy/blob/e8fe59f83843ca619264e8d4757bc99fcde93fa9/modules/dhcp_isc/configuration_loader.rb#L28

Note you can use `container.get_dependency(:database)` there to pass the DB connection.

It's then a regular class: https://github.com/theforeman/smart-proxy/blob/develop/modules/dhcp_common/free_ips.rb

As you scale up, this may work better.